eth/client_test.go

```diff
@@ -0,0 +1,153 @@
+			DelegatedStake: big.NewInt(3),
+		},
+		{
+			Address:        ethcommon.HexToAddress("ddd"),
+			DelegatedStake: big.NewInt(2),
+		},
+		{
+			Address:        ethcommon.HexToAddress("eee"),
+			DelegatedStake: big.NewInt(1),
+		},
+	}
+
+	// transcoder is not in pool and doesn't have enough stake to join the pool
+	hints := simulateTranscoderPoolUpdate(ethcommon.HexToAddress("fff"), big.NewInt(1), copyTranscoders(transcoders), true)
+	assert.Equal(hints, lpTypes.StakingHints{})
+
+	// transcoder is not in pool and pool is full, transcoder joins the list, current tail is evicted
+	pool := copyTranscoders(transcoders)
+	hints = simulateTranscoderPoolUpdate(ethcommon.HexToAddress("fff"), big.NewInt(2), pool, true)
+	assert.Equal(hints, lpTypes.StakingHints{
+		PosPrev: ethcommon.HexToAddress("ddd"),
+	})
+	assert.NotEqual(pool[len(pool)-1].Address, ethcommon.HexToAddress("eee"))
+
+	// transcoder is not in pool and pool is not full
+	// transcoder takes last spot
+	hints = simulateTranscoderPoolUpdate(ethcommon.HexToAddress("fff"), big.NewInt(1), copyTranscoders(transcoders), false)
+	assert.Equal(hints, lpTypes.StakingHints{
+		PosPrev: ethcommon.HexToAddress("eee"),
+	})
+
+	// transcoder is not in pool and pool is not full
+	// transcoder takes first spot
+	hints = simulateTranscoderPoolUpdate(ethcommon.HexToAddress("fff"), big.NewInt(100), copyTranscoders(transcoders), false)
+	assert.Equal(hints, lpTypes.StakingHints{
+		PosNext: ethcommon.HexToAddress("aaa"),
+	})
+
+	// transcoder is not in pool and pool is not full
+	// transcoder takes second spot
+	hints = simulateTranscoderPoolUpdate(ethcommon.HexToAddress("fff"), big.NewInt(5), copyTranscoders(transcoders), false)
+	assert.Equal(hints, lpTypes.StakingHints{
+		PosPrev: ethcommon.HexToAddress("aaa"),
+		PosNext: ethcommon.HexToAddress("bbb"),
+	})
+
+	// last transcoder's stake increases but remains in the last spot
+	hints = simulateTranscoderPoolUpdate(ethcommon.HexToAddress("eee"), big.NewInt(2), copyTranscoders(transcoders), false)
+	assert.Equal(hints, lpTypes.StakingHints{
+		PosPrev: ethcommon.HexToAddress("ddd"),
+	})
+
+	// transcoder is in pool and moves up one from the last spot
+	pool = copyTranscoders(transcoders)
+	hints = simulateTranscoderPoolUpdate(ethcommon.HexToAddress("eee"), big.NewInt(3), pool, false)
+	assert.Equal(hints, lpTypes.StakingHints{
+		PosPrev: ethcommon.HexToAddress("ccc"),
+		PosNext: ethcommon.HexToAddress("ddd"),
+	})
+	assert.Equal(ethcommon.HexToAddress("ddd"), pool[len(pool)-1].Address)
+
+	// transcoder is in pool and takes the first spot
+	hints = simulateTranscoderPoolUpdate(ethcommon.HexToAddress("ccc"), big.NewInt(6), copyTranscoders(transcoders), false)
+	assert.Equal(hints, lpTypes.StakingHints{
+		PosNext: ethcommon.HexToAddress("aaa"),
+	})
+
+	// transcoder decreases stake , takes the last spt
+	hints = simulateTranscoderPoolUpdate(ethcommon.HexToAddress("aaa"), big.NewInt(0), copyTranscoders(transcoders), false)
```

Comment:
Wouldn't that result in that transcoder taking the one before last spot because it was already in the list, had a higher stake originally and we traverse in descending order? If it were a new transcoder not in the list yet, it would take the last spot. 

Agreed that using '0'  is not representative of actual on-chain behaviour, I could bump everything up by 1.

Reply:
Actually scratch the above,  I think it should indeed take the last spot because it would get removed first and then re-added. This seems to be inconsistent with the way we sort the slice.

Reply:
Thanks your comment made me change that test and observed it failing, which uncovered that bug.

We now mimic the on-chain behaviour of `updateKey()` 

1. If the transcoder is in the pool remove it 
2. Append the transcoder 
3. Sort the list 
4. Evict the last if  list is full 
5. Get the hints 

Also added a test that confirms we no longer encounter this bug

 84c2b0c9

Reply:
I think this bug might actually exist in LivepeerJS as well https://github.com/livepeer/livepeerjs/blob/0e5edcc097a972f5a4221055617a31b48f912b5d/packages/explorer-2.0/lib/utils.tsx#L355

Reply:
Just to clarify:

Suppose we have transcoders T1 and T2. In the pool, if transcoders have the same stake, both the ordering T1 -> T2 and the ordering T2 -> T1 are valid based on the [validInsertPosition()](https://github.com/livepeer/protocol/blob/826af1124f645a819b54eb0500ddc31f0d2af755/contracts/libraries/SortedDoublyLL.sol#L250) logic in the SortedDoublyLL library right now. To illustrate:

**Scenario 1**

T1 is inserted into the pool with a hint where `prev = T2`. T1 would be inserted in the pool after T2.

**Scenario 2**

T1 is inserted into the pool without a hint. We would descend the list of transcoders in the pool. When the next transcoder in the pool is T2 we would see that T1's stake is >= the prev transcoder's stake and is = to T2's stake. This is also a valid insert position for T1.